rfcs/0007-pass-spec.md

```diff
@@ -0,0 +1,463 @@
+
+#### Pointer capture
+
+When a user starts a long press on a widget, the widget can "capture" the pointer.
+
+Pointer capture has a few implications:
+
+- When a widget has captured a pointer, all events from that pointer will be sent to the widget, even if the pointer isn't in the widget's hitbox.
+Conversely, no other widget can get events from the pointer.
+- The "hovered" status of other widgets won't be updated even if the pointer is over them.
+The hovered status of the capturing widget will be updated, meaning a widget that captured a pointer can still lose the "hovered" status.
+- The pointer's cursor icon will be updated as if the pointer stayed over the capturing widget.
+- If the widget loses pointer capture for some reason (e.g. the pointer is disconnected), the Widget will get a `PointerLeave` event.
+
+Masonry should guarantee that pointers can only be captured by one widget at a time.
+Masonry should force the widget to lose pointer capture when some events occur; not just MouseLeave, but also `Tab` being pressed, the window losing focus, the widget being disabled, etc.
+
+Examples of use cases for pointer capture include selecting text, dragging a slider, or long-pressing a button.
+
+#### Focus
+
+Focus marks whether a widget receives text events.
+
+To give a simple example, when you click a textbox, the textbox gets focus: anything you type on your keyboard will be sent to that textbox.
+
+Focus can be changed with the tab key, or by clicking on a widget, both which Masonry automatically handles.
+Widgets can also set custom focus behavior.
+
+Note that widgets without text-edition capabilities such as buttons and checkboxes can also get focus.
+For instance, pressing space when a button is focused will trigger that button.
+
+There are two types of focus: active and inactive focus.
+Active focus is the default one; inactive focus is when the window your app runs in has lost focus itself.
+
+In that case, we still mark the widget as focused, but with a different color to signal that e.g. typing on the keyboard won't actually affect it.
+
+
+## Implementation strategy
+
+### More explicitly document passes in the code
+
+Comments and method names in `RenderRoot` and `WidgetPod` should explicitly call out passes where they happen.
+
+The "User-facing explanation" section should be added to the crate's documentation, and in-code documentation should refer to it.
+
+### Rename WidgetCtx to MutateCtx
+
+Since the MUTATE pass is becoming a documented part of the code, MutateCtx would be a clearer name for that context type.
+
+### Change how widgets are added
+
+The WidgetAdded event should be removed for `Lifecycle`, and the `WidgetPodInner` type should be removed.
+
+Instead, widgets should be created and added to the widget tree as a single atomic operation.
+To keep the WidgetPod logic simple and avoid too many corner cases, this is only allowed inside the MUTATE pass.
```

Comment:
I wonder how that will look like in practice? 
I.e. with different constructors for different kind of widgets.
so is something like this still valid (more pseudocode, this is wrong of course):

```rust
let button = Button::new(..);
// later e.g. for a flex
flex_widget.insert_flex_child_pod(button);
```

Or how should that roughly look like?
Is the constructor for a widget just a data container, and gets properly constructed when adding that child to whatever widget it contains?

(But the general idea sounds good)

Reply:
Yeah, we're still trying to figure that out.

See [this discussion on zulip](https://xi.zulipchat.com/#narrow/stream/317477-masonry/topic/Implementing.20the.20new.20pass.20system/near/457380951).